keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,462 @@
+    * Feature gate for `NonGracefulFailover` is disabled, feature is not active.
+    * Feature gate for `NonGracefulFailover` is enabled. Add `out-of-service` taint after node is shutdown:
+      * Verify workloads are moved to another node successfully.
+      * Verify the `out-of-service` taint is removed after the shutdown node is cleaned up.
+* Add stress and scale tests before moving from beta to GA.
+
+We also plan to test this with different version Skews.
+
+### Graduation Criteria
+
+This KEP will be treated as a new feature, and will be introduced with a new feature gate, `NonGracefulFailover`.
+
+This enhancement will go through the following maturity levels: alpha, beta and stable.
+
+#### Alpha
+
+* Initial feature implementation in kube-controller-manager.
+* Add basic unit tests.
+
+#### Alpha -> Beta Graduation
+
+* Gather feedback from developers and surveys.
+* Unit tests and e2e tests outlined in design proposal implemented.
+* Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+* Feature is deployed in production and have gone through at least one Kubernetes upgrade.
+* More rigorous forms of testing—e.g., downgrade tests and scalability tests.
+* Allowing time for feedback.
+
+### Upgrade / Downgrade Strategy
+
+once enabled by default, the upgrades should be smooth as the default behaviour is off by default. when downgrading/rollbacking with this behaviour enabled, there should be no issues as the control plane with the logic is still there.
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+
+-->
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [ ] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: NonGracefulFailover
+    - Components depending on the feature gate: kube-controller-manager
```

Comment:
it surprises me that there isn't work to do on the kubelet side for handling when a kubelet that was previously thought lost forever reconnects.

Reply:
This KEP does not propose any changes in Kubelet. The proposed changes are in the Pod CG Controller and Attach/detach controller.
Since the pods are forcefully deleted from the shutdown node and VolumeAttachments are also deleted before the node comes up and accept new pods to be scheduled on it again, we don't need to make changes in Kubelet.
The current KEP does rely on the user to apply the taint when the node is down and remove the taint when the node is recovered. In the future, we'll make this more automatic.

Reply:
One potential problem I see with this is user applies the out-of-service taint, the pod is rescheduled and the volume attachments are removed, and later the pod is recovered. Correct me if I'm wrong, but when the node comes back and the user subsequently removes the out-of-service taint, if the node was not rebooted during that period (perhaps it only had loss of connectivity), theire might be remaining mounts or paths on the node for the volumes that were not cleaned up because NodeUnpublishVolume and NodeUnstageVolume were not called. If you subsequently force another reschedule the pods that would result in in trying to create a replace pod back on the original node, it may fail to mount the volumes, depending on how the CSI driver views these existing mounts. 
So is a reboot of the node required during the period of application of the out-of-service taint to removal of the out-of-service taint?

Reply:
i agree, there is no change in the kubelet.

this is really about the standard convention for the taint, and making the volume controller respect it.

Reply:
shutdown case means the node should be rebooted when it comes back. For other non-recoverable cases like hardware failure, I think OS reboot will be part of repair? We can make this point clear in the doc.

Reply:
Added a note in the KEP.